src/main/java/io/supertokens/totp/Totp.java

```diff
@@ -13,47 +33,143 @@
 import io.supertokens.totp.exceptions.LimitReachedException;
 
 public class Totp {
+    private static String generateSecret() throws Exception {
+        final String TOTP_ALGORITHM = "HmacSHA1";
+
+        try {
+            final KeyGenerator keyGenerator = KeyGenerator.getInstance(TOTP_ALGORITHM);
+            keyGenerator.init(160); // 160 bits = 20 bytes
 
-    public static String generateSecret() {
-        return "XXXX";
+            // FIXME: Should return base32 or base16
+            // Return base64 string of the secret key:
+            return Base64.getEncoder().encodeToString(keyGenerator.generateKey().getEncoded());
+        } catch (NoSuchAlgorithmException e) {
+            throw new Exception("TOTP algorithm not found");
+        }
     }
 
-    public static boolean checkCode(TOTPDevice device, String code) {
-        if (code.startsWith("XXXX")) {
-            return true;
+    private static boolean checkCode(TOTPDevice device, String code) {
+        final TimeBasedOneTimePasswordGenerator totp = new TimeBasedOneTimePasswordGenerator();
+
+        byte[] keyBytes = Base64.getDecoder().decode(device.secretKey);
+        Key key = new SecretKeySpec(keyBytes, "HmacSHA1");
+
+        final int period = device.period;
+        final int skew = device.skew;
+
+        // Check if code is valid for any of the time periods in the skew:
+        for (int i = -skew; i <= skew; i++) {
+            try {
+                // FIXME: Where is this using % on the period?
+                if (totp.generateOneTimePasswordString(key, Instant.now().plusSeconds(i * period)).equals(code)) {
+                    return true;
+                }
+            } catch (InvalidKeyException e) {
+                return false;
+            }
         }
+
         return false;
     }
 
-    public static String createDevice(Main main, String userId, String deviceName, int skew, int period)
-            throws StorageQueryException, DeviceAlreadyExistsException {
+    /**
+     * Replicates TOTP code is generated by apps like Google Authenticator and Authy
+     */
+    @TestOnly
+    public static String generateTotpCodeForDevice(TOTPDevice device) throws InvalidKeyException {
+        final TimeBasedOneTimePasswordGenerator totp = new TimeBasedOneTimePasswordGenerator();
+
+        byte[] keyBytes = Base64.getDecoder().decode(device.secretKey);
+        Key key = new SecretKeySpec(keyBytes, "HmacSHA1");
+
+        return totp.generateOneTimePasswordString(key, Instant.now());
+    }
```

Comment:
this doesn't seem to be calling any function interally to this file - move this to a test file.

Reply:
good point. done